### DISCUSSION TITLE
Enforcing SACC file/object as an input to `build_likelihood` method

### DISCUSSION THREAD

**Discussion by arthurmloureiro** [LSSTDESC/firecrown#421] [Ideas]
Hi!

So from building the blinding infrastructure in a way that is "likelihood agnostic", I noticed one potential issue which I cannot protect the user from and that is having a hardcoded path to a sacc file in the likelihood.

For example, in the new [firecrown tutorial documents](https://firecrown.readthedocs.io/en/latest/_static/two_point_generators.html) we have an example where the SACC file is somewhat harcoded to the `build_likelihood()` function:

```python
SACC_FILENAME = os.path.expandvars( \
    "${FIRECROWN_DIR}/examples/srd_sn/srd-y1-converted.sacc" \
    )

def build_likelihood(_):
    snia_stats = sn.Supernova(sacc_tracer="sn_ddf_sample")
    likelihood = ConstGaussian(statistics=[snia_stats])
    sacc_data = sacc.Sacc.load_fits(SACC_FILENAME)
    likelihood.read(sacc_data)
    # This simple likelihood has no associated ModelingTools
    return likelihood, None
```
 
Meanwhile, [TXPipe deals with this in a different way](https://github.com/LSSTDESC/TXPipe/blob/a849e2d65e844aeed1499b0d0919a6f8c7bbfc29/txpipe/utils/theory_model.py#L13):
```python
def build_likelihood(build_parameters):
    """
    This is a generic 3x2pt theory model, but with only
    bias as a systematic.

    """
    sacc_data = build_parameters['sacc_data']
[...]
```

For the blinding infrastructure, since we want the user to be able to blind a sacc file from the command line (as well as to import the library and use it in TXPipe) there is a potential issue if the user provides a likelihood for the blinding that works in the way that the first example above works because we use the likelihood method `get_data_vector()` to obtain the DV to be blinded. If we have a `def build_likelihood(_)` with a hardcoded sacc file, the blinding will pass the user sacc file to this likelihood module which is ignored, using the hardcoded sacc instead.

I don't seem to have a way to impose that any likelihoods passed to the blinding infrastructure should take a sacc file as input to the `build_likelihood` function.

In [my likelihood example for the blinding](https://github.com/LSSTDESC/Blinding/blob/533319c8ca5ea054fb4733a3ef05687add35198e/examples/cosmic_shear/cosmicshear_likelihood.py#L17C1-L42C37), I even have to implement some hacking to protect from giving the SACC file in different ways:
```python
def build_likelihood(build_parameters):
    """
    Create a firecrown likelihood for a cosmic shear analysis.
    
    Parameters
    ----------
    build_parameters : dict
        Dictionary of parameters to be used in the likelihood.

    Returns
    -------
    likelihood : firecrown.likelihood.likelihood.Likelihood
        A firecrown likelihood object.
    """

    try:
        sacc_data = build_parameters['sacc_data']
    except TypeError:
        sacc_data = build_parameters.data['sacc_data']

    # This is a trick TXPipe uses to pass a sacc file or object
    # to the likelihood. We need to check if the input is a string
    if isinstance(sacc_data, (str, pathlib.Path)):
        sacc_data = sacc.Sacc.load_fits(sacc_data)
    else:
        sacc_data = sacc_data.copy()
```

Personally, I think that we should discourage the user from creating likelihoods with hardcoded data-vectors. Is that something others agree? that maybe we should impose that the `build_likelihood` method must receive an input which contains either a sacc object or a path to a sacc file?

Cheers!

**Comment by arthurmloureiro** [comment]
Tagging @marcpaterno @vitenti and @tilmantroester here for the discussion 😄

**Comment by tilmantroester** [comment]
Related: #394
Maybe changing the example to something that's not using a hardcoded sacc file might be a partial but quick fix.

**Reply by arthurmloureiro** [reply]
Agreed! We had people in the Forecasting Workshop struggling to understand how not to give a hardcoded sacc to the cosmosis firecrown likelihood.

**Comment by arthurmloureiro** [comment]
I really think we should enforce that sacc files cannot be hardcoded into the likelihood.

**Comment by marcpaterno** [comment]
The factory mechanism introduced into Firecrown v1.8 allows the (name of the)  SACC file to be passed as a parameter.

**Reply by arthurmloureiro** [reply]
Hi @marcpaterno !

So the issue is not being able to pass the name of the sacc as a parameter. The issue is that the current structure allows for a sacc file path to be hardcoded inside the firecrown likelihood. In that case, the user can pass a path to a sacc file externally that is simply not going to be used. There's no way to test for this properly (unless something really changed ... I haven't tested it in a while).

**Reply by tilmantroester** [reply]
I don't think it's possible to guarantee that. But a solution would be for firecrown to store a reference to the sacc file that was being read when the likelihood was built, which could then be queried by the blinding library to check if it matches the sacc file that is supposed to be blinded.

**Reply by marcpaterno** [reply]
@arthurmloureiro Is this issue still relevant for the blinding project? If so, what should we save? Our first proposal would be:

1. A likelihood could record a hash identifying the SACC object used to initialize it (if SACC objects are hashable; we're not sure about that), and if the likelihood was created using a SACC object (they do not have to be).
2. If the likelihood is modified after creation, the hash of the SACC object would be cleared (because it does not then reflect the state of the likelihood after configuration from only the SACC object).
3. When generating mock data (e.g. for forecasting) the user is responsible for providing a hash.
4. The blinding code can access the hash from the likelihood, and if it is not None, can verify that the right SACC was used.

Would this be helpful for the blinding?

**Reply by arthurmloureiro** [reply]
Hi @marcpaterno any way of ensuring the firecrown likelihood does not contain a hardcoded sacc file in it would be vital not only for blinding but for DESC as a whole.

The sacc file should be an input to the cobaya/cosmosis/numcosmos config for the firecrown likelihoods. Most of the examples have hardcoded sacc files in the "likelihood.py" and that can lead to catastrophic errors (running the likelihood on the unblinded sacc, for example)

**Reply by tilmantroester** [reply]
There can be legitimate uses for hardcoding a sacc file in a likelihood, e.g. in cases where the likelihood is specific to this sacc file.

**Reply by arthurmloureiro** [reply]
> There can be legitimate uses for hardcoding a sacc file in a likelihood, e.g. in cases where the likelihood is specific to this sacc file. 

Hi @tilmantroester I think in codes like cosmosis, for example, even in these cases we have the config pointing to the data.

These cases should be avoided, in my fair opinion. There's no way to ensure this is not happening for the blinding and that's a bigger problem than allowing the user to have bad habits.

**Reply by tilmantroester** [reply]
I agree that having the hardcoded paths in the examples is not ideal but I don't think it's technically feasible to enforce the absence of hardcoded paths. The best and only option is to have the blinding code check that the file that was loaded into the likelihood is indeed the one that it's supposed to.

**Reply by arthurmloureiro** [reply]
You can enforce that that a sacc path needs to be an input to the `build_likelihood` function.
I think this pretty much stops the sacc being hardcoded?

> The best and only option is to have the blinding code check that the file that was loaded into the likelihood is indeed the one that it's supposed to.
I can try that! I think I tried it back in the days and failed, but I can try again ! Thanks for the idea 😄

**Reply by tilmantroester** [reply]
>  I think this pretty much stops the sacc being hardcoded?

Someone could still load a different sacc file within the `build_likelihood` function, either intentionally or by accident. I don't see how you could prevent that.

>  I can try that! I think I tried it back in the days and failed, but I can try again ! Thanks for the idea 😄

I'm not sure that functionality exists yet but @marcpaterno 's proposal above would address this.
